src/Libraries/Microsoft.Extensions.Telemetry/Logging/ExtendedLogger.cs

```diff
@@ -0,0 +1,396 @@
+            LegacyPath<TState>(logLevel, eventId, state, exception, formatter);
+        }
+    }
+
+    private static LoggerMessageState MessageProps
+    {
+        get
+        {
+            var props = _loggerMessageState;
+            if (props == null)
+            {
+                props = new();
+                _loggerMessageState = props;
+            }
+            else
+            {
+                _ = props.TryReset();
+            }
+
+            return props;
+        }
+    }
+
+    private LogMethodHelper MethodHelper
+    {
+        get
+        {
+            var helper = _logMethodHelper;
+            if (helper == null)
+            {
+                helper = new();
+                _logMethodHelper = helper;
+            }
+            else
+            {
+                _ = helper.TryReset();
+            }
+
+            helper.AddRange(_staticProperties);
+            return helper;
+        }
+    }
+
+    private PropertyJoiner Joiner
+    {
+        get
+        {
+            var joiner = _joiner;
+            if (joiner == null)
+            {
+                joiner = new()
+                {
+                    StaticProperties = _staticProperties
+                };
+
+                _joiner = joiner;
+            }
+
+            return joiner;
+        }
+    }
+
+    private PropertyBag Bag
+    {
+        get
+        {
+            var bag = _bag;
+            if (bag == null)
+            {
+                bag = new();
+                _bag = bag;
+            }
+            else
+            {
+                bag.Clear();
+            }
+
+            bag.StaticProperties = _staticProperties;
+            return bag;
+        }
+    }
+
+    private void ModernPath(LogLevel logLevel, EventId eventId, LoggerMessageState? properties, Exception? exception, Func<LoggerMessageState, Exception?, string> formatter)
+    {
+        // presume warning/error/critical levels are enabled, to avoid the call to IsEnabled
+        if (logLevel < LogLevel.Warning)
+        {
+            if (!IsEnabled(logLevel))
+            {
+                return;
+            }
+        }
+
+        properties ??= MessageProps;
+
+        try
+        {
+            // redact
+            if (_redactorProvider != null)
+            {
+                foreach (var cp in properties.ClassifiedProperties)
+                {
+                    properties.AddProperty(cp.Name, _redactorProvider.GetRedactor(cp.Classification).Redact(cp.Value));
+                }
+            }
+
+            // enrich
+            foreach (var enricher in _enrichers)
+            {
+                enricher.Enrich(properties.EnrichmentPropertyBag);
+            }
+
+            // one last dedicated bit of enrichment
+            if (exception != null && _includeStackTraces)
+            {
+                properties.AddProperty(ExceptionStackTrace, GetExceptionStackTrace(exception));
+            }
+
+            var joiner = Joiner;
```

Comment:
Hmmm. Someone calling a log method from within the formatter function seems to be intentionally asking for trouble. But I can imagine it happening purely through composition in some error case for example.

So how about I use CompareExchange to get the stuff from TLS and set it to null during execution, and then put it back when I'm done. If some other code executes reentrantly, it wil see null values in the TLS and will allocate new objects in that case.

Either that, or I go back to object pools.

Reply:
> So how about I use CompareExchange to get the stuff from TLS and set it to null during execution

The CompareExchange should be unneeded because the value is already threadlocal. A non-atomic null write should work fine. TLS reads/writes have more overhead than a standard read/write (I think 5-10ns vs. fraction of ns) so if you are trying to squeeze out performance one thing you can do is:

```C#
class ThreadLocalPool
{
    //whatever objects you want TLS singletons for
    Type1 Object1 { get; set; } = new Object1();
    Type2 Object2 { get; set; } = new Object2();
    ...
}


[ThreadStatic]
static ThreadLocalPool? _pool;

void DoSomethingThatNeedsObj1()
{
    ThreadLocalPool pool = _pool; // a single TLS read in the normal case
    if(pool == null)
    {
        pool = _pool = new ThreadLocalPool();
    }
    // these reads and writes are standard fast field accesses
    Type1 obj1 = poolObject1 ?? new Object1();
    pool.Object1 = null;
    // use obj1
    pool.Object1 = obj1;
}
```